src/H5FDsubfiling/H5FDsubfiling.c

```diff
@@ -2581,138 +2043,913 @@ H5FD__subfiling_ctl(H5FD_t *_file, uint64_t op_code, uint64_t flags, const void
+             */
+            if (cur_file_offset % stripe_size)
+                cur_max_num_subfiles++;
+        }
+
+        loc_max_num_subfiles = MAX((size_t)cur_max_num_subfiles, loc_max_num_subfiles);
+    }
+
+    *max_iovec_depth  = loc_max_iovec_depth;
+    *max_num_subfiles = loc_max_num_subfiles;
+}
+
+/*-------------------------------------------------------------------------
+ * Function:    translate_io_req_to_iovec
+ *
+ * Purpose:     Helper routine to perform the translation between an I/O
+ *              request (type, addr, size, buf tuple) and a set of I/O
+ *              vectors that spans all the subfiles touched by that I/O
+ *              request. Once finished, this function will have generated
+ *              at most `iovec_count` sets of I/O vectors, each containing
+ *              `iovec_len` elements, but a smaller number of I/O vector
+ *              sets could be generated, depending on the I/O request.
+ *
+ *              sf_context (IN)
+ *                - the subfiling context for the file
+ *
+ *              iovec_idx (IN)
+ *                - the index into `io_types`, `io_addrs`, `io_sizes` and
+ *                  `io_bufs` where this function should begin filling in
+ *                  the I/O vectors
+ *
+ *              iovec_len (IN)
+ *                - the number of elements in each I/O vector generated
+ *
+ *              iovec_count (IN)
+ *                - the maximum number of I/O vectors to be generated, as
+ *                  calculated in generate_io_vectors()
+ *
+ *              type (IN)
+ *                - the memory type to use for each component of the I/O
+ *                  vectors generated
+ *
+ *              addr (IN)
+ *                - the starting file offset used to generate the I/O
+ *                  vectors
+ *
+ *              io_size (IN)
+ *                - the size of the I/O to the given file offset, which is
+ *                  used when generating the I/O vectors
+ *
+ *              io_buf (IN)
+ *                - the I/O buffer to be partitioned up while generating
+ *                  the I/O vectors
+ *
+ *              io_type (IN)
+ *                - the type of I/O being performed (IO_TYPE_WRITE or
+ *                  IO_TYPE_READ)
+ *
+ *              io_types (OUT)
+ *                - pointer to the memory types I/O vector to populate
+ *
+ *              io_addrs (OUT)
+ *                - pointer to the file offsets I/O vector to populate
+ *
+ *              io_sizes (OUT)
+ *                - pointer to the I/O sizes I/O vector to populate
+ *
+ *              io_bufs (OUT)
+ *                - pointer to the I/O buffers I/O vector to populate
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+static herr_t
+translate_io_req_to_iovec(subfiling_context_t *sf_context, size_t iovec_idx, size_t iovec_len,
```

Comment:
Much of the logic in this function is unchanged from the previous function it was in (`init_indep_io`). The main difference is that the buffer indexing had to change since the I/O vectors used for the final I/O requests are populated directly here rather than generating arrays that the vectors would have later been populated with.